File: scripts/mcp_smoke.py
#!/usr/bin/env python3
"""Smoke test for the linux-x11-sandbox MCP stdio server."""

import json
import subprocess
import sys


def main() -> int:
    proc = subprocess.Popen(
        ["./target/release/linux-x11-sandbox"],
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )

    def call(method: str, params: dict, req_id: int) -> dict:
        msg = (
            json.dumps(
                {"jsonrpc": "2.0", "id": req_id, "method": method, "params": params}
            )
            + "\n"
        )
        proc.stdin.write(msg)  # type: ignore[union-attr]
        proc.stdin.flush()  # type: ignore[union-attr]
        line = proc.stdout.readline()  # type: ignore[union-attr]
        return json.loads(line)

    init = call(
        "initialize",
        {
            "protocolVersion": "2024-11-05",
            "capabilities": {},
            "clientInfo": {"name": "smoke", "version": "0.1.0"},
        },
        1,
    )
    assert init["result"]["serverInfo"]["name"] == "linux-x11-sandbox", init

    tools = call("tools/list", {}, 2)
    tool_names = {t["name"] for t in tools["result"]["tools"]}
    required = {
        "lxs_display_create",
        "lxs_display_destroy",
        "lxs_display_info",
        "lxs_app_launch",
        "lxs_app_terminate",
        "lxs_app_list",
        "lxs_input_click",
        "lxs_input_move",
        "lxs_input_scroll",
        "lxs_input_type",
        "lxs_input_key",
        "lxs_capture_screenshot",
        "lxs_capture_region",
        "lxs_state_window",
        "lxs_state_tree",
        "lxs_state_element_bounds",
        "lxs_perform_action",
    }
    missing = required - tool_names
    assert not missing, f"missing tools: {missing}"

    display = call(
        "tools/call", {"name": "lxs_display_create", "arguments": {}}, 3
    )
    display_id = display["result"]["display_id"]
    assert display["result"]["display"].startswith(":"), display

    info = call(
        "tools/call",
        {"name": "lxs_display_info", "arguments": {"display_id": display_id}},
        4,
    )
    assert info["result"]["width"] > 0, info
    assert info["result"]["height"] > 0, info

    shot = call(
        "tools/call",
        {"name": "lxs_capture_screenshot", "arguments": {"display_id": display_id}},
        5,
    )
    assert shot["result"]["mimeType"] == "image/png", shot
    assert len(shot["result"]["data"]) > 0, shot

    destroyed = call(
        "tools/call",
        {"name": "lxs_display_destroy", "arguments": {"display_id": display_id}},
        6,
    )
    assert destroyed["result"]["success"], destroyed

    proc.terminate()
    try:
        proc.wait(timeout=5)
    except subprocess.TimeoutExpired:
        proc.kill()

    print("mcp smoke test passed")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: lxs/mcp/tests/mcp_smoke.rs
use std::io::{BufRead, BufReader, Write};
use std::process::{Child, Command, Stdio};

use serde_json::{json, Value};

struct McpSession {
    child: Child,
    reader: BufReader<std::process::ChildStdout>,
    writer: std::process::ChildStdin,
    next_id: i64,
}

impl McpSession {
    fn new() -> Self {
        let bin = env!("CARGO_BIN_EXE_linux-x11-sandbox");
        let mut child = Command::new(bin)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()
            .expect("failed to spawn linux-x11-sandbox");

        let reader = BufReader::new(child.stdout.take().unwrap());
        let writer = child.stdin.take().unwrap();

        Self {
            child,
            reader,
            writer,
            next_id: 1,
        }
    }

    fn call(&mut self, method: &str, params: Value) -> Value {
        let id = self.next_id;
        self.next_id += 1;

        let req = json!({
            "jsonrpc": "2.0",
            "id": id,
            "method": method,
            "params": params,
        });

        let msg = req.to_string() + "\n";
        self.writer.write_all(msg.as_bytes()).unwrap();
        self.writer.flush().unwrap();

        let mut line = String::new();
        self.reader.read_line(&mut line).unwrap();
        serde_json::from_str(&line).expect("invalid json response")
    }
}

impl Drop for McpSession {
    fn drop(&mut self) {
        let _ = self.child.kill();
        let _ = self.child.wait();
    }
}

#[test]
fn mcp_smoke() {
    let mut s = McpSession::new();

    let init = s.call(
        "initialize",
        json!({
            "protocolVersion": "2024-11-05",
            "capabilities": {},
            "clientInfo": { "name": "smoke", "version": "0.1.0" },
        }),
    );
    assert_eq!(init["result"]["serverInfo"]["name"], "linux-x11-sandbox");

    let tools = s.call("tools/list", json!({}));
    let names: Vec<String> = tools["result"]["tools"]
        .as_array()
        .unwrap()
        .iter()
        .map(|t| t["name"].as_str().unwrap().to_string())
        .collect();

    let required = [
        "lxs_display_create",
        "lxs_display_destroy",
        "lxs_display_info",
        "lxs_app_launch",
        "lxs_app_terminate",
        "lxs_app_list",
        "lxs_input_click",
        "lxs_input_move",
        "lxs_input_scroll",
        "lxs_input_type",
        "lxs_input_key",
        "lxs_capture_screenshot",
        "lxs_capture_region",
        "lxs_state_window",
        "lxs_state_tree",
        "lxs_state_element_bounds",
        "lxs_perform_action",
    ];
    for name in required {
        assert!(names.contains(&name.to_string()), "missing tool {}", name);
    }

    let display = s.call(
        "tools/call",
        json!({"name": "lxs_display_create", "arguments": {}}),
    );
    let display_id = display["result"]["display_id"]
        .as_str()
        .unwrap()
        .to_string();
    assert!(display["result"]["display"]
        .as_str()
        .unwrap()
        .starts_with(':'));

    let info = s.call(
        "tools/call",
        json!({
            "name": "lxs_display_info",
            "arguments": { "display_id": display_id },
        }),
    );
    assert!(info["result"]["width"].as_u64().unwrap() > 0);
    assert!(info["result"]["height"].as_u64().unwrap() > 0);

    let shot = s.call(
        "tools/call",
        json!({
            "name": "lxs_capture_screenshot",
            "arguments": { "display_id": display_id },
        }),
    );
    assert_eq!(shot["result"]["mimeType"], "image/png");
    assert!(!shot["result"]["data"].as_str().unwrap().is_empty());

    let destroyed = s.call(
        "tools/call",
        json!({
            "name": "lxs_display_destroy",
            "arguments": { "display_id": display_id },
        }),
    );
    assert_eq!(destroyed["result"]["success"], true);
}
